src/internal_modules/roc_sndio/wav_sink.cpp

```diff
@@ -0,0 +1,268 @@
+/*
+ * Copyright (c) 2023 Roc Streaming authors
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#include "roc_sndio/wav_sink.h"
+#include "roc_core/endian_ops.h"
+#include "roc_core/log.h"
+#include "roc_core/panic.h"
+#include "roc_sndio/backend_map.h"
+
+namespace roc {
+namespace sndio {
+
+WavSink::WavSink(core::IArena& arena, const Config& config)
+    : output_file_(NULL)
+    , header_(config.sample_spec.num_channels(), config.sample_spec.sample_rate(), 32)
+    , buffer_(arena)
+    , buffer_size_(0)
+    , valid_(false) {
+    BackendMap::instance();
+
+    if (config.sample_spec.num_channels() == 0) {
+        roc_log(LogError, "wav sink: # of channels is zero");
+        return;
+    }
+
+    if (config.latency != 0) {
+        roc_log(LogError, "wav sink: setting io latency not supported by wav backend");
+        return;
+    }
+
+    frame_length_ = config.frame_length;
+
+    if (frame_length_ == 0) {
+        roc_log(LogError, "wav sink: frame length is zero");
+        return;
+    }
+
+    valid_ = true;
+}
+
+WavSink::~WavSink() {
+    close_();
+}
+
+bool WavSink::is_valid() const {
+    return valid_;
+}
+
+bool WavSink::open(const char* path) {
+    roc_panic_if(!valid_);
+
+    roc_log(LogDebug, "wav sink: opening: path=%s", path);
+
+    if (output_file_ != NULL) {
+        roc_panic("wav sink: can't call open() more than once");
+    }
+
+    if (!open_(path)) {
+        return false;
+    }
+
+    if (!setup_buffer_()) {
+        return false;
+    }
+
+    return true;
+}
+
+DeviceType WavSink::type() const {
+    return DeviceType_Sink;
+}
+
+DeviceState WavSink::state() const {
+    return DeviceState_Active;
+}
+
+void WavSink::pause() {
+    // no-op
+}
+
+bool WavSink::resume() {
+    return true;
+}
+
+bool WavSink::restart() {
+    return true;
+}
+
+audio::SampleSpec WavSink::sample_spec() const {
+    roc_panic_if(!valid_);
+
+    if (!output_file_) {
+        roc_panic("wav sink: sample_spec(): non-open output file or device");
+    }
+
+    if (header_.num_channels() == 1) {
+        return audio::SampleSpec(size_t(header_.sample_rate()),
+                                 audio::ChanLayout_Surround, audio::ChanOrder_Smpte,
+                                 audio::ChanMask_Surround_Mono);
+    }
+
+    if (header_.num_channels() == 2) {
+        return audio::SampleSpec(size_t(header_.sample_rate()),
+                                 audio::ChanLayout_Surround, audio::ChanOrder_Smpte,
+                                 audio::ChanMask_Surround_Stereo);
+    }
```

Comment:
I've chosen to use default constructor for `ChannelSet` and then set what's needed, instead of using other constructor, which would require using a mask. It could've been temporarily initialized with a mono mask and then set, but I think the code is cleaner if it's done this way:
```c++
audio::ChannelSet channel_set;
    channel_set.set_layout(audio::ChanLayout_Surround);
    channel_set.set_order(audio::ChanOrder_Smpte);
    channel_set.set_channel_range(0, header_.num_channels() - 1, true);
```

Reply:
Makes sense